imxrt-hal/src/adc.rs

```diff
@@ -0,0 +1,275 @@
+//!
+//! let reading: u16 = adc1.read(&mut a1).unwrap();
+//!```
+
+use crate::ccm;
+use crate::iomuxc::adc::{prepare_adc_pin, Pin, ADC1, ADC2};
+use crate::iomuxc::{adc, consts::Unsigned};
+use crate::ral;
+use core::marker::PhantomData;
+use embedded_hal::adc::{Channel, OneShot};
+
+/// The clock input for an ADC
+#[allow(non_camel_case_types)]
+pub enum ClockSelect {
+    IPG,   // IPG clock
+    IPG_2, // IPG clock / 2
+    ADACK, // Asynchronous clock
+}
+
+/// How much to divide the clock input
+pub enum ClockDivision {
+    Div1, // Input clock / 1
+    Div2, // Input clock / 2
+    Div4, // Input clock / 4
+    Div8, // Input clock / 8
+}
+
+impl ClockSelect {
+    pub fn default() -> Self {
+        ClockSelect::ADACK
+    }
+}
+
+impl ClockDivision {
+    pub fn default() -> Self {
+        ClockDivision::Div2
+    }
+}
+
+/// Conversion speeds done by clock cycles
+pub enum ConversionSpeed {
+    Slow,     // 25 ADC clock cycles (24 on imxrt102x)
+    Medium,   // 17 ADC clock cycles (16 on imxrt102x)
+    Fast,     // 9 ADC clock cycles (8 on imxrt102x)
+    VeryFast, // 3 ADC clock cycles (2 on imxrt102x)
+}
+
+/// Denotes how much hardware averaging to do
+pub enum AveragingCount {
+    Avg1,
+    Avg4,
+    Avg8,
+    Avg16,
+    Avg32,
+}
+
+// Specifies the resolution the ADC
+pub enum ResolutionBits {
+    Res8,
+    Res10,
+    Res12,
+}
+
+/// A pin representing an analog input for a particular ADC
+pub struct AnalogInput<ADCx, P> {
+    _module: PhantomData<ADCx>,
+    _pin: P,
+}
+
+impl<P, ADCx> Channel<ADCx> for AnalogInput<ADCx, P>
+where
+    P: Pin<ADCx>,
+    ADCx: adc::ADC,
+{
+    type ID = u16;
+
+    fn channel() -> Self::ID {
+        <P as Pin<ADCx>>::Input::U16
+    }
+}
+
+impl<P, ADCx> AnalogInput<ADCx, P>
+where
+    P: Pin<ADCx>,
+    ADCx: adc::ADC,
+{
+    /// Creates a new analog input pin
+    pub fn new(mut pin: P, _adc: &ADC<ADCx>) -> Self {
+        prepare_adc_pin(&mut pin);
+        Self {
+            _module: PhantomData,
+            _pin: pin,
+        }
+    }
+}
+
+pub struct ADC<ADCx> {
+    _module: PhantomData<ADCx>,
+    reg: ral::adc::Instance,
+}
+
+impl<ADCx> ADC<ADCx> {
+    fn new(reg: ral::adc::Instance) -> Self {
+        let inst = Self {
+            _module: PhantomData,
+            reg,
+        };
+
+        inst.set_resolution(ResolutionBits::Res10);
+        inst.set_low_power_mode(false);
+
+        // Calibrate w/ slow settings initially
+        inst.set_averaging(AveragingCount::Avg32);
+        inst.set_conversion_speed(ConversionSpeed::Slow);
+        inst.calibrate();
+
+        // Set to default of 4 hardware averages & medium conversion speed
+        inst.set_averaging(AveragingCount::Avg4);
+        inst.set_conversion_speed(ConversionSpeed::Medium);
+
+        inst
+    }
+
+    /// Sets the resolution that analog reads return, in bits
+    pub fn set_resolution(&self, bits: ResolutionBits) {
```

Comment:
This kind of pertains to the above conversation about interrupt safety: should these `set_*` methods take a `&mut self` receiver, rather than a `&self`? Same for `calibrate()`? They're mutating something owned by the ADC.

The `Oneshot` method already takes a `&mut self` receiver. Once everything it signaling mutability, anyone who wants to modify an `ADC` across execution contexts will need to wrap it in a mutex, or their favorite interrupt-free construct, to synchronize it across execution contexts.

Reply:
I think that mostly solves my concerns if you want to expose the set_ style configuration functions, they should take a &mut self, good call.

It sort of leaves it up to the user to figure out how to synchronize and make it safe. In the future perhaps a configure(&self, cfg: Configuration) could be provided which does either a disable, config, enable or an atomic spin loop waiting for interrupts to stop doing things perhaps.